### PR TITLE
Fix DDP bug in single process multiple device use cases

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36503 [WIP] Fix DDP bug in single process multiple device use cases**

Differential Revision: [D21179274](https://our.internmc.facebook.com/intern/diff/D21179274)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -314,6 +314,20 @@ def _ddp_init_helper(self):
...
+
+        def parameters(m, recurse=True):
+            def model_parameters(m):
+                ps = m._former_parameters.values() \
+                    if hasattr(m, "_former_parameters") \
+                    else m.parameters(recurse=False)
+                for p in ps:
+                    yield p
+
+            ms = m.modules() if recurse else [m]
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:326]: what is this for? looks like it is not used?

**Review by zhaojuanmao — APPROVED** [review]

**Review by zhaojuanmao — COMMENTED** [review]
sorry, accidentally clicked approve

File: test/distributed/test_c10d.py [hunk 2]
```diff
@@ -1871,6 +1871,89 @@ def forward(self, x):
         return F.softmax(x, dim=1).to(dev0)
 
 
+@unittest.skipIf(TEST_WITH_TSAN, "TSAN is not fork-safe since we're forking in a multi-threaded environment")
+class DistributedDataParallelSingleProcessTest(TestCase):
```
Comment (zhaojuanmao) [inline test/distributed/test_c10d.py:1875]: it has distributeddataparallel test in test_distributed.py,  do you think whether it is better  to put tests in test_distributed.py?

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (mrshenli) [reply torch/nn/parallel/distributed.py:326]: my bad will update.

File: test/distributed/test_c10d.py (hunk 2 above)
Comment (mrshenli) [reply test/distributed/test_c10d.py:1875]: Sure, let me move it. Would I be correct if I assume we plan to move all DDP related tests to `test_distributed.py`?

Comment (zhaojuanmao) [reply test/distributed/test_c10d.py:1875]: yeah, I feel it is better to put ddp related tests in test_distributed.py, how do you think?

Comment (mrshenli) [reply test/distributed/test_c10d.py:1875]: Hey @zhaojuanmao, I compared `test_distributed.py` and `test_c10d.py`. If we plan to make them `test_ddp.py` and `test_c10d.py` in the future, I agree that this test better fits in `test_ddp.py`. But for now, it looks it fits more in `test_c10d.py`, as `DistributedDataParallelSingleProcessTest` is the single-process version of `DistributedDataParallelTest ` in `test_c10d.py` immediately after this test. And currently, a considerable part of `test_distributed.py` are testing APIs on c10d backend. Shall we keep this test in here for now, and we can certainly move it when we refactor the test file structures.

**Review by zhaojuanmao — APPROVED** [review]
sounds good, I will think about how to refactor the test structures as well.

thanks for fixing it.

**Pull request closed without merging**